From 229ed45778621556c24d0a7770518f0551f7c1ad Mon Sep 17 00:00:00 2001
From: Ilya Dryomov <idryomov@gmail.com>
Date: Sun, 16 Aug 2026 18:39:03 +0200
Subject: [PATCH] librbd/cache/pwl/ssd: avoid dispatch_deferred_writes() vs
 ~AbstractWriteLog() race

In append_op_log_entries(), dispatch_deferred_writes() is called after
append_ops() returns.  This can result in a race between AbstractWriteLog
destructor and dispatch_deferred_writes() because all I/Os may complete
and append_ctx callback may run allowing AbstractWriteLog::shut_down()
to indicate completion before dispatch_deferred_writes() is called.

This race is problematic even if there aren't any deferred writes
(i.e. m_deferred_ios list is empty): a deadlock may pop up because
m_thread_pool is stopped by the destructor under the same lock that
dispatch_deferred_writes() needs to perform the initial check on
m_deferred_ios.

Call dispatch_deferred_writes() from append_ops() before the I/O batch
is submitted, similar to what was done previously in commit d83a0f6db8ff
("librbd/cache/pwl/ssd: avoid corrupting m_first_free_entry").  Since
dispatch_deferred_writes() appears to be called to act on a decremented
m_bytes_allocated, this doesn't change anything for it.  The ops.size()
condition has always been redundant as append_op_log_entries() asserts
on it and ops list remains constant.

Fixes: https://tracker.ceph.com/issues/77894
Signed-off-by: Ilya Dryomov <idryomov@gmail.com>
---
 src/librbd/cache/pwl/ssd/WriteLog.cc | 5 +----
 1 file changed, 1 insertion(+), 4 deletions(-)

diff --git a/src/librbd/cache/pwl/ssd/WriteLog.cc b/src/librbd/cache/pwl/ssd/WriteLog.cc
index 42f04692cc0d..4f7d2f260e0e 100644
--- a/src/librbd/cache/pwl/ssd/WriteLog.cc
+++ b/src/librbd/cache/pwl/ssd/WriteLog.cc
@@ -520,10 +520,6 @@ void WriteLog<I>::append_op_log_entries(GenericLogOperations &ops) {
     });
   // Append logs and update first_free_update
   append_ops(ops, append_ctx, new_first_free_entry);
-
-  if (ops.size()) {
-    this->dispatch_deferred_writes();
-  }
 }
 
 template <typename I>
@@ -898,6 +894,7 @@ void WriteLog<I>::append_ops(GenericLogOperations &ops, Context *ctx,
     m_first_free_entry = *new_first_free_entry;
     m_bytes_allocated -= bytes_to_free;
   }
+  this->dispatch_deferred_writes();
 
   bdev->aio_submit(&aio->ioc);
 }
